Add CredScan to the PR builds

erdembayar opened the pull request:
<!-- DO NOT MODIFY OR DELETE THIS TEMPLATE. IT IS USED IN AUTOMATION. -->
## Bug

<!-- Search https://github.com/NuGet/Home/issues, and create one if you can't find a suitable issue. -->
<!-- Paste the full link, like https://github.com/nuget/home/issues/1000. GitHub will render is neatly. -->
Fixes: https://github.com/NuGet/Client.Engineering/issues/954

Regression? Last working version:

## Description
Add cred scan into PR builds, creates asset file then analyze it, if detects any credential included in PR repo ~~upon build then creates issues DevDiv area and send notification email to hotseat~~ fails immediately. 
If you want to see credscanner work in action [check](https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=5310371&view=logs&j=428ae689-995e-5327-66dd-018ba6766d0a&t=4b8463db-e99b-5830-a106-c33198e92386&l=12), here it fails due detecting 2 violations from [commit#26c9223ccbce0546070ab0a620187be9dd410479](https://github.com/NuGet/NuGet.Client/commit/26c9223ccbce0546070ab0a620187be9dd410479).
On observation is it's ignoring default parameters values, it's same for both my local and CI build.
For completeness there is another tool called "Guardian" but uses older version of CredScanner, and I find it less sensitive, it's quite different.

## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
  - [ ] Automated tests added
  - **OR**
  <!-- Describe why you haven't added automation. -->
  - [ ] Test exception
  - **OR**
  - [x] N/A <!-- Infrastructure, documentation etc. -->

- **Documentation**
  <!-- Please link the PR/issue if appropriate -->
  - [ ] Documentation PR or issue filled
  - **OR**
  - [x] N/A

erdembayar committed:
Start adding CredScan into PR.

Add resources.

Move cred scan into build.

Stop creating TSA issues for PR builds.

Enable logginc CredScan result.

Fail

Fail on error.

Download artifact.

kartheekp-ms commented on eng/pipelines/templates/Build_and_UnitTest.yml:
```diff
@@ -26,6 +26,64 @@ steps:
+    CredScan: true
+    ToolLogsNotFoundAction: "Standard"
+
+- task: TSAUpload@1
```
I think `TSAUpload task` is not required for PR builds. We should fail the build if issues are detected.

erdembayar replied:
~~Thank you for your review. How do I fail the build from `CredScan@2` task after `SdtReport@1` is published with `PublishSecurityAnalysisLogs`.  If I just use `continueOnError: "false` with CredScan then it can't publish the report.~~
Please review now.

erdembayar committed:
Print current loc

No need download

1

See artifact folder.

Print env

Add back

erdembayar committed:
Print

erdembayar committed:
Fail if detect credential in PR.

erdembayar committed:
Add CredScan suppression file.

erdembayar committed:
Move CredScanSuppression file.

erdembayar committed:
Revert back unneeded changes.

erdembayar commented on .config/CredScanSuppressions.json:
```diff
@@ -0,0 +1,13 @@
+        {
+            "_justification": "Unit tests use dummy input data.",
+            "file": [
+                "test/NuGet.Core.Tests/NuGet.Protocol.Tests/Plugins/Messages/GetCredentialsResponseTests.cs",
```
It looks there 2 ways of suppression, [inline or suppression file](https://strikecommunity.azurewebsites.net/articles/8148/handling-false-positives-and-suppressions.html).

erdembayar replied:
Without suppression file it'll [fail](https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=5284233&view=logs&j=d80b2849-30d4-52dd-74cd-d6536ba98fe0&t=1827c9b6-09e3-515b-9a63-74df1158657e&l=15).

erdembayar replied:
Below are actual offending lines:

https://github.com/NuGet/NuGet.Client/edit/dev/docs/cross-platform-debugging.md#L66

https://github.com/NuGet/NuGet.Client/blob/966b1b21eec28834365a8604744dab682c8326ea/test/NuGet.Core.Tests/NuGet.Protocol.Tests/Plugins/Messages/GetCredentialsResponseTests.cs#L72-L73
https://github.com/NuGet/NuGet.Client/blob/966b1b21eec28834365a8604744dab682c8326ea/test/NuGet.Core.Tests/NuGet.Protocol.Tests/Plugins/Messages/GetCredentialsResponseTests.cs#L90-L91
https://github.com/NuGet/NuGet.Client/blob/966b1b21eec28834365a8604744dab682c8326ea/test/NuGet.Core.Tests/NuGet.Protocol.Tests/Plugins/Messages/GetCredentialsResponseTests.cs#L109-L112

https://github.com/NuGet/NuGet.Client/blob/966b1b21eec28834365a8604744dab682c8326ea/test/NuGet.Core.Tests/NuGet.Protocol.Tests/Plugins/Messages/SetCredentialsRequestTests.cs#L30 (CredScanner detect as line 37)
https://github.com/NuGet/NuGet.Client/blob/966b1b21eec28834365a8604744dab682c8326ea/test/NuGet.Core.Tests/NuGet.Protocol.Tests/Plugins/Messages/SetCredentialsRequestTests.cs#L49
https://github.com/NuGet/NuGet.Client/blob/966b1b21eec28834365a8604744dab682c8326ea/test/NuGet.Core.Tests/NuGet.Protocol.Tests/Plugins/Messages/SetCredentialsRequestTests.cs#L72

erdembayar commented:
@NuGet/nuget-client 
Please review.

zivkan commented on eng/pipelines/templates/Build_and_UnitTest.yml:
```diff
@@ -26,6 +26,76 @@ steps:
+    CredScan: true
+    ToolLogsNotFoundAction: "Standard"
+
+- task: PublishSecurityAnalysisLogs@2
```
Do we need to publish the logs for PR builds?  If it's going to email the team, and maybe create a work item for the team to fix, this is too much for code that hasn't been merged into dev.

erdembayar replied:
Ok. Removed.

zivkan commented on test/EndToEnd/tests/GetPackageTest.ps1:
```diff
@@ -181,6 +181,8 @@ function Test-GetPackageForProjectReturnsCorrectPackages2 {
 
 function Test-GetPackageForFSharpProjectReturnsCorrectPackages {
+    [SkipTest('https://github.com/NuGet/Home/issues/11291')]
+    param()
```
It's better to do this in a separate PR beause
1. It can get merged faster, as it won't get slowed down by other comments in this PR
2. If this PR introduces problems into the product, or build, it can be reverted, without re-enabling a flaky test.

zivkan commented on eng/pipelines/templates/Build_and_UnitTest.yml:
```diff
@@ -26,6 +26,76 @@ steps:
     inlineScript: |
       Get-ChildItem Env: | Sort-Object Name | Format-Table -Wrap -AutoSize
 
+- task: CredScan@2
```
Build_and_UnitTest is on the hot path, therefore making this job slower is going to make the entire pipeline slower. It would be better to create a new job that runs in parallel. Afterwards, we can move the `dotnet-format`, and any other check that can be done independently, to the new job.

This will also avoid CredScan running on both the RTM and non-RTM jobs.

erdembayar replied:
I placed on separate pipeline now.

zivkan commented on eng/pipelines/templates/Build_and_UnitTest.yml:
```diff
@@ -26,6 +26,76 @@ steps:
       Get-ChildItem Env: | Sort-Object Name | Format-Table -Wrap -AutoSize
 
+- task: CredScan@2
+  continueOnError: "false"
```
`false` is the default for continueOnError, you don't need to explicitly set it here, or on line 36 below.

erdembayar replied:
Removed.

erdembayar commented:
@zivkan @kartheekp-ms @dominoFire @heng-liu 
Please review.

erdembayar committed:
Address PR comment

erdembayar committed:
Move CredScan to separate pipeline.

erdembayar committed:
Don't pass variables.

erdembayar committed:
Add back commented code.

kartheekp-ms commented:
Can you please rebase your branch again? I see lot of other commits added to this PR.

erdembayar committed:
Merge branch 'dev' into dev-eryondon-954

erdembayar commented:
> Can you please rebase your branch again? I see lot of other commits added to this PR.

@zivkan @kartheekp-ms 
Thank you for your reviews. I addressed PR comments, please review again.

erdembayar commented:
@NuGet/nuget-client 
Please review.

donnie-msft commented on eng/pipelines/templates/Scan_Code.yml:
```diff
@@ -0,0 +1,38 @@
+        }
+
+        if ($credScannerXml.DEFECTS.DEFECT -ne $null) {
+          Write-Host "CredentialScanner found following credential in this PR, please remove them from PR and use git squash/drop to remove commits with credential from git history then do force push:" 
```
Are you sure squashing is adequate? My understanding is the commits still exist on the remote.
Anyone who obtained the commit id or github URL from a PR/etc can still access the original commit.

![image](https://user-images.githubusercontent.com/49205731/137248125-87c64896-bce8-4d92-83d6-36e095a25d6f.png)

https://github.com/NuGet/NuGet.Client/commit/5c96ea582e2cf75622ee771530799d1b255b6d2d

erdembayar replied:
You're right, yet can't think of better idea. I'm open for any other suggestions.
Even deleting actual github commit might not be enough,  not sure how often and what rate 3rd parties aggregate github.  For example [here's](https://www.gitmemory.com/NuGet/NuGet.Client) snapshot of nuget.

donnie-msft replied:
There's an article about it, mentioning tools I've not used myself:  eg, `git filter-repo `.
[Removing sensitive data from a repository](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/removing-sensitive-data-from-a-repository)

donnie-msft replied:
Would it be enough to just say here "revoke this credential immediately"? Whether the commits are deleted isn't as important as revocation, right? Ideally, yes, we also scrub the history, but as you say it's not catching 100% of the spill.

erdembayar replied:
> There's an article about it, mentioning tools I've not used myself: eg, `git filter-repo `. [Removing sensitive data from a repository](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/removing-sensitive-data-from-a-repository)

Most likely that tool is not enough in worst case scenario, someone or 3rd party bots already get the copy.

erdembayar replied:
> Would it be enough to just say here "revoke this credential immediately"? Whether the commits are deleted isn't as important as revocation, right? Ideally, yes, we also scrub the history, but as you say it's not catching 100% of the spill.

Sure, but I want to merge while I got green build. May I reword it in another PR? Reworded text would be `CredentialScanner found following credential in this PR, please revoke this credential immediately and remove them from PR and use git squash/drop to remove commits with credential from git history then do force push:`

donnie-msft replied:
Sure, follow-up is fine! Thanks

donnie-msft replied:
> and remove them from PR 

> and use git squash/drop to remove commits with credential from git history then do force push

...still think one or both of these parts can be omitted.

donnie-msft replied:
Why not just:

> CredentialScanner found following credential in this PR, please revoke this credential immediately.

erdembayar replied:
Created follow up issue https://github.com/NuGet/Client.Engineering/issues/1250

zivkan commented on eng/pipelines/templates/pipeline.yml:
```diff
@@ -64,6 +64,17 @@ stages:
     - template: Initialize_Build.yml
 
+- stage: Scan_Code
+  displayName: Pre-Checks before start build
```
This isn't running before the build starts, so the display name isn't right. But I want to move the `dotnet-format` check into this pipeline (meaning Scan_Code probably also isn't a good name). I know names are hard, but just an FYI to think about in future PRs to have the highest quality PRs for the team to review.

erdembayar replied:
> This isn't running before the build starts, so the display name isn't right. But I want to move the `dotnet-format` check into this pipeline (meaning Scan_Code probably also isn't a good name). I know names are hard, but just an FYI to think about in future PRs to have the highest quality PRs for the team to review.

Created follow up issue https://github.com/NuGet/Client.Engineering/issues/1250.
